design-proposals/primary_network_binding_for_seamless_migration.md

```diff
@@ -0,0 +1,459 @@
+The community has already shown interest in this direction with an
+active [contribution](https://github.com/kubevirt/kubevirt/pull/12671).
+
+A plugin solves the live-migration support indicator (as it is part of
+the [plugin API](https://kubevirt.io/user-guide/network/network_binding_plugins/#migration))
+but on the other hand introduces a challenge with the DHCP servers and
+data sharing between the CNI and virt-launcher (in the core, this is
+solved by the virt-handler persisting the data on a file for the
+virt-launcher to consume).
+
+IPv6 support remains similar to the core binding version.
+
+There is one additional implementation challenge which comes with the
+plugin: The creation of the tap device by the CNI requires to correctly
+consider the SELinux labeling in advance because tap devices cannot
+be relabeled after creation.
+
+From a maintenance point of view, the core is untouched and kept simple
+while the burden is added to the plugin. For gradual adoption, the
+plugin seems a better choice. But if the scenario is expected to be
+commonly used, it may add redundant overhead.
+As a plugin, it can be developed independently to the Kubevirt release
+cycle on a different repo.
+
+#### Macvtap Binding
+
+```mermaid
+graph LR;
+    subgraph virt-launcher
+    VM <--> macvtap0
+    macvtap0 <--> eth0
+    end
+    subgraph Node
+    eth0 <--> veth123
+    end
+```
+
+The macvtap binding is an alternative to the bridge binding with very
+similar pros and cons.
+The [macvtap device](https://virt.kernelnewbies.org/MacVTap) is attached
+to the pod interface in a passthru mode.
+
+> *Warning*: Requires PoC to confirm it works when attaching to a veth
+> interface and multiple mac addresses are active in the guest.
+
+Deployed as a binding plugin with no option to self-serve IPAM to the VM.
+It is dependent on the network-provider to be able and serve IP
+addresses to the guest.
+
+Compared to the bridge binding (as a plugin), it is simpler and with
+probably better performance.
+In addition, macvtap devices have a device representation and can be
+relabeled (SELinux) post creation.
+
+> *Warning*: Relabeling of a macvtap device needs PoC confirmation.
+
+As a plugin, it can be developed independently to the Kubevirt release
+cycle on a different repo.
+
+#### Other L2 Bindings
+
+The bridge and macvtap bindings have one behavior in common: They
+forward traffic based on L2 (ethernet frames) information only.
+
+Similar bindings may be invented to forward traffic using other technologies.
+Some may be more efficient than others with different complexities but
+with a very similar pros and cons as seen on the previous bridge &
+macvtap bindings, e.g.:
+
+- Passing the IP & route configuration from the pod interface to the VM
+  guest interface is left to the network provider (e.g. DHCPv4, DHCPv6, RA).
+
+Some of the suggested L2 alternatives are:
+
+- OVS bridge
+- eBPF
+- Tap device directly connected to the node, passed into the pod for
```

Comment:
In the past we asked about this and looks like tap device get disconnected when moved to the network namespace, although I don't remeber the details.